### PR TITLE
Ecrf File Generation Fix

### PR DESCRIPTION
**Pull request by isaaccodekill**
- Include third party data
- Include Emission factor details

### What change is being made?
Enhance ECRF file generation by incorporating additional data attributes for accurate inventory reporting and ensure correct handling of third-party data sources in the processing logic.

### Why are these changes being made?
This change addresses the need for more comprehensive data handling in emissions inventory management by including additional attributes such as `datasetName`, `datasourceName`, and `dataQuality` in inventory data processing. This is to accommodate third-party data sources effectively, ensuring that the inventory correctly reflects third-party data characteristics and custom emission factors. The adjustment also corrects a logical error in how zero values were handled in direct measuring formulae, preventing erroneous data validations.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: app/src/components/Modals/activity-modal/activity-modal-body.tsx [hunk 1]
```diff
@@ -140,6 +133,8 @@ const ActivityModalBody = ({
       );
       const emissionFactorType = emissionFactorTypeValue;
       if (emissionFactorType === "custom") {
+        setValue("activity.emissionFactorReference", "custom emission factors");
```
Comment (lemilonkh) [inline app/src/components/Modals/activity-modal/activity-modal-body.tsx:136]: Do we need to use i18n here?

File: app/src/components/Modals/activity-modal/activity-modal-body.tsx [hunk 2]
```diff
@@ -140,6 +133,8 @@ const ActivityModalBody = ({
...
+        setValue("activity.emissionFactorName", "custom");
```
Comment (lemilonkh) [inline app/src/components/Modals/activity-modal/activity-modal-body.tsx:137]: I18n?

File: app/src/backend/ECRFDownloadService.ts [hunk 3]
```diff
@@ -103,67 +103,102 @@ export default class ECRFDownloadService {
...
+                activityValue.metadata?.emissionFactorName,
+              emission_factor_description:
+                activityValue.metadata?.emissionFactorTypeReference,
+              activity_unit: activityUnit,
+              methodology: t(methodologyDescription),
+              activity_data_quality: dataQuality,
+              activity_data_source: dataSource,
+              total_co2e: toDecimal(activityValue.co2eq as bigint)
+                ?.div(new Decimal("1e3"))
+                .toNumber(),
```
Comment (lemilonkh) [inline app/src/backend/ECRFDownloadService.ts:197]: If you call .toNumber() at the end, you lose a lot of precision. Use strings where necessary to transport decimals across API/ DB boundaries.

File: app/src/hooks/activity-value-form/use-emission-factors.ts [hunk 4]
```diff
@@ -6,19 +6,24 @@ import { EmissionsFactorResponse } from "@/util/types";
...
+  name: string;
+  reference: string;
+  gasValuesByGas: {
+    [gas: string]: {
+      gasValues: Record<string, any>[];
+    };
+  };
+}
+
+[];
```
[the remark this answers is not in the document]
Comment (lemilonkh) [reply app/src/hooks/activity-value-form/use-emission-factors.ts:20]: Remove this line please

File: app/src/backend/ECRFDownloadService.ts (hunk 3 above)
Comment (isaaccodekill) [reply app/src/backend/ECRFDownloadService.ts:197]: Will correct in a future clean up

File: app/src/backend/formulas.ts [hunk 5]
```diff
@@ -194,18 +194,48 @@ function convertDataToDefaultUnit(
...
+  // Check if all gas amounts are zero
+  const allZero = gases.every((gas) => gas.amount.equals(0));
+
+  if (allZero) {
+    throw new createHttpError.BadRequest(
+      "Direct measure requires a non zero gas amount",
+    );
+  }
+
+  // TODO: Save amounts to GasValue entries or perform further processing
```
Comment (lemilonkh) [inline app/src/backend/formulas.ts:237]: Is this still required?

File: app/src/components/Tabs/Activity/direct-measure-table.tsx [hunk 6]
```diff
@@ -130,13 +130,23 @@ const DirectMeasureTable: FC<DirectMeasureTableProps> = ({
                   </Text>
                 </Td>
                 <Td isNumeric isTruncated>
-                  {convertKgToTonnes(activity?.activityData?.co2_amount)}
+                  {/*Direct measure entries are collected in tonnes by default*/}
```
Comment (lemilonkh) [inline app/src/components/Tabs/Activity/direct-measure-table.tsx:133]: I think our default is to save all gas amounts in kg actually? At least that was the intent...

**Review by lemilonkh — APPROVED** [review]
Thanks 👍🏼✨

**Pull request merged**